File: 1475. Final Prices With a Special Discount in a Shop/Solution.cpp
#include <stack>
#include <vector>
using namespace std;

class Solution
{
public:
    vector<int> finalPrices(vector<int> &prices)
    {
        stack<int> stk;
        for (size_t i = 0; i < prices.size(); i++) {
            while (!stk.empty() && prices[stk.top()] >= prices[i]) {
                prices[stk.top()] -= prices[i];
                stk.pop();
            }
            stk.emplace(i);
        }
        return prices;
    }
};

File: 1475. Final Prices With a Special Discount in a Shop/Solution.py
from typing import List


class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        stack = []
        for i, p in enumerate(prices):
            while stack and prices[stack[-1]] >= p:
                prices[stack.pop()] -= p
            stack.append(i)
        return prices
